xarray/backends/zarr.py

```diff
@@ -421,21 +446,38 @@ def maybe_decode_store(store, lock=False):
 
     # auto chunking needs to be here and not in ZarrStore because variable
     # chunks do not survive decode_cf
-    if auto_chunk:
-        # adapted from Dataset.Chunk()
-        def maybe_chunk(name, var):
-            from dask.base import tokenize
-            chunks = var.encoding.get('chunks')
-            if (var.ndim > 0) and (chunks is not None):
-                # does this cause any data to be read?
-                token2 = tokenize(name, var._data)
-                name2 = 'zarr-%s' % token2
-                return var.chunk(chunks, name=name2, lock=None)
-            else:
-                return var
-
-        variables = OrderedDict([(k, maybe_chunk(k, v))
-                                 for k, v in ds.variables.items()])
-        return ds._replace_vars_and_dims(variables)
-    else:
+    # return trivial case
+    if not chunks:
         return ds
+
+    # adapted from Dataset.Chunk()
+    if isinstance(chunks, int):
+        chunks = dict.fromkeys(ds.dims, chunks)
+
+    def selkeys(dict_, keys):
+        if dict_ is None:
+            return None
+        return dict((d, dict_[d]) for d in keys if d in dict_)
+
+    def maybe_chunk(name, var, chunks):
+        from dask.base import tokenize
+
+        if chunks == 'auto':
+            chunks = var.encoding.get('chunks')
+        else:
+            chunks = selkeys(chunks, var.dims)
```

Comment:
No particular thoughts from me.  You would have to look at the `.chunks` attribute of the dask array and compare it to the chunking of the zarr data.  You might also consider erring or automatically rechunking if they don't align well.  

If you go this route then you might want to look at `dask.array.core.normalize_chunks`, and set both `previous_chunks` to the zarr array chunks and set `limit` to some nice byte size.

```python
normalize_chunks('auto', shape=..., limit='100MiB', previous_chunks=zarr_dataset.chunks)
```

I haven't actually tried that, and it's been a while since I dealt with the auto-rechunking code, so no guarantees on the above.  I encourage others to investigate it as an option.

Reply:
@jcrist might have different thoughts though.  You could also ask @jakirkham, who probably knows more here.

Reply:
That does sound interesting. A warning sounds better than automatic re-chunking, which could be frustrating, especially if the data is small enough to re-chunk with relatively little issue. I'm at a summer school with no signal and little wifi, so I'll look at this when I get back!

Reply:
Any thoughts on the warning?